src/Core/KeyManagement/UserKey/Implementations/RotateUserAccountkeysCommand.cs

```diff
@@ -97,95 +97,45 @@ public async Task<IdentityResult> RotateUserAccountKeysAsync(User user, RotateUs
     {
-        // A V2 user has a signature key pair, and their user key is COSE
-        // The user key cannot be directly checked here; but the items encrypted with it can be checked.
-        ArgumentNullException.ThrowIfNull(user);
-        var hasSignatureKeyPair = await _userSignatureKeyPairRepository.GetByUserIdAsync(user.Id) != null;
-        var isPrivateKeyEncryptionV2 = GetEncryptionType(user.PrivateKey) == EncryptionType.XChaCha20Poly1305_B64;
-
-        // Valid v2 user
-        if (hasSignatureKeyPair && isPrivateKeyEncryptionV2)
-        {
-            return true;
-        }
-
-        // Valid v1 user
-        if (!hasSignatureKeyPair && !isPrivateKeyEncryptionV2)
-        {
-            return false;
-        }
-
-        throw new InvalidOperationException("User is in an invalid state for key rotation. User has a signature key pair, but the private key is not in v2 format, or vice versa.");
-    }
+        ValidateV2Encryption(model);
+        await ValidateVerifyingKeyUnchanged(model, user);
 
-    public async Task ValidateRotationModelSignatureKeyPairForV2User(RotateUserAccountKeysData model, User user)
-    {
-        var currentSignatureKeyPair = await _userSignatureKeyPairRepository.GetByUserIdAsync(user.Id);
-        if (model.AccountKeys == null || model.AccountKeys.SignatureKeyPairData == null)
-        {
-            throw new InvalidOperationException("The provided signing key data is null, but the user already has signing keys.");
-        }
-        if (model.AccountKeys.SignatureKeyPairData.VerifyingKey != currentSignatureKeyPair.VerifyingKey)
-        {
-            throw new InvalidOperationException("The provided verifying key does not match the user's current verifying key.");
-        }
-        if (string.IsNullOrEmpty(model.AccountKeys.PublicKeyEncryptionKeyPairData?.SignedPublicKey))
-        {
-            throw new InvalidOperationException("No signed public key provided, but the user already has a signature key pair.");
-        }
-        if (GetEncryptionType(model.AccountKeys.SignatureKeyPairData.WrappedSigningKey) != EncryptionType.XChaCha20Poly1305_B64)
-        {
-            throw new InvalidOperationException("The provided signing key data is not wrapped with XChaCha20-Poly1305.");
-        }
+        saveEncryptedDataActions.Add(_userSignatureKeyPairRepository.UpdateForKeyRotation(user.Id, model.AccountKeys.SignatureKeyPairData));
+        user.SignedPublicKey = model.AccountKeys.PublicKeyEncryptionKeyPairData.SignedPublicKey;
     }
 
-    public void ValidateRotationModelSignatureKeyPairForV1UserAndUpgradeToV2(RotateUserAccountKeysData model, User user, List<UpdateEncryptedDataForKeyRotation> saveEncryptedDataActions)
+    public void UpgradeV1ToV2KeysAsync(RotateUserAccountKeysData model, User user, List<UpdateEncryptedDataForKeyRotation> saveEncryptedDataActions)
```

Comment:
Technically this does not need to be `Async` suffixed, since it's not async function
```suggestion
    public void UpgradeV1ToV2Keys(RotateUserAccountKeysData model, User user, List<UpdateEncryptedDataForKeyRotation> saveEncryptedDataActions)
```